### DISCUSSION TITLE
How does one sequence resource deployments with Flux?

### DISCUSSION THREAD

**Discussion by vijayansarathy** [fluxcd/flux2#2183] [General] [answered]
Hi,
Does Flux V2 provide a mechanism to orchestrate multiple deployments in a specific ordered sequence to ensure certain resources are healthy before subsequent resources are synced/reconciled. Also, the ability to execute specific actions after a resource is synced/reconciled to the cluster, like, for example, pausing for X seconds after A is deployed before we started syncing B.  ArgoCD provides sync waves and hooks that allow one to do this. I was wondering what the equivalent mechanisms are in Flux.
Thanks,
Viji

**Comment by stefanprodan** [comment, marked as answer]
You can use dependsOn and health checks to define the apply order: https://fluxcd.io/docs/components/kustomize/kustomization/#kustomization-dependencies

**Reply by Snixells** [reply]
I could not find this here so i thought it's a good idea to post it: The current Flux Version supports the "DependsOn" for HelmReleases too: https://fluxcd.io/flux/components/helm/helmreleases/

**Comment by youwalther65** [comment]
@stefanprodan  What's about having a Kustomization resource which is dependend on a HelmRelease? My example is a Karpenter installation as a HelmRelease. Once that is finished I want to install so-called provisioner which are YAML. But the depend on the Karpenter release to be ready and provide the CRD/API. Using something like:

$ cat karpenter-kustomization.yaml
---
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: karpenter-provisioner
  namespace: karpenter
spec:
  interval: 10m0s
  path: ./karpenter/Kustomize
  prune: true
  dependsOn:
    - name: karpenter-release
      namespace: karpenter
  sourceRef:
    kind: GitRepository
    name: flux-apps
    namespace: flux-system

or
...
  dependsOn:
    - name: HelmRelease/karpenter-release
      namespace: karpenter
...

$ flux get kustomizations -n karpenter
NAME                    READY   MESSAGE                                                                                                                         REVISION        SUSPENDED
karpenter-provisioner   False   unable to get 'karpenter/karpenter-release' dependency: Kustomization.kustomize.toolkit.fluxcd.io "karpenter-release" not found                 False

does not work. Do you suggest using healthChecks and point to a deployment created by the previous HelmRelease?

**Reply by stefanprodan** [reply]
> Do you suggest using healthChecks and point to a deployment created by the previous HelmRelease?

You can set `wait: true` in the Kustomization that applies the Karpenter HelmRelease, then all the other Kustomizations that depend on it, will execute after the Karpenter CRDs and workloads are ready.

**Comment by mibollma** [comment]
I have a similar, probably pretty common case and I'm a bit confused about the recommended way to handle those dependencies between helm releases and other resources.

A simplified view of my dependencies look like this:

- SealedSecret helm chart
- SomeSealedSecret // <-- depends on SealedSecret helm chart
- SomeOtherChart // <-- references SomeSealedSecret

@stefanprodan So if I understand correctly the recommended way to resolve this is to split those elements into 3 Kustomizations and settings **dependsOn** as well as **wait** appropriately?

**Reply by youwalther65** [reply]
I am confused as well. I used a HelmRelase resource for the Karpenter Helm provisioning and Kustomization for Karpenter provisioner YAML (depends on Karpenter API) . How do I "move" the HelmRelease to a Kustomization

**Reply by stefanprodan** [reply]
Check out https://github.com/fluxcd/flux2-kustomize-helm-example This shows how you can define two layers for your cluster, one is `infrastructure` (CRDs, controllers, cluster addons, etc) and the other one `apps`. The `apps` Kustomizations depends on `infrastructure`. By setting `wait: true` in the `infrastructure` Kustomizations, Flux will wait for all Helm releases to be installed then it proceeds to `apps`.

**Reply by youwalther65** [reply]
Thank you @stefanprodan I am aware of this already and it works. But hoped for a simpler solution. But now I think of the Karpenter HelmRelease and the dependend Karpenter provisioner as a kind of infrastructure and "app" ;-) So I have to split-up similar use cases accordingly f- fro example AWS EBS CSI drover in "infrastructure" and "derived" storageclass in "app"

**Reply by frank-bee** [reply]
@stefanprodan I'm working on a multicluster setup with flux currently where I have a similar question.
 I want to have a re-usable infrastructure setup for each cluster. Idea: Using a Flux Kustomization referencing a Flux Gitrepo, specific version or branch.
The re-usable infrastructure consists of several components ( some optional). These components have to be installed in an order.
Should I used nested Kustomizations for these comp. (with depends on), HelmReleases, oder even a combination of both? 

Mainly the components are Helmreleases ( ISTIO , Ext. Secrets Controller,...) , that's why it would be nice to use Flux Helmreleases instead of Flux Kustomizations. But there might by some yamls as well. E.g. After 1) Ext. secrets controller is installed , 2)  a SecretProvider CRDS has to be instantiated and 3) afterwards other charts will be installed....

So the depends-on cannot combine Helmreleases and Kustomizations? ( This would be possible in Argo CD , where Helmrelease is just a "subtype" of an Application)

**Reply by stefanprodan** [reply]
A HelmRelease is just a subset of a Flux Kustomization. Here is an example https://github.com/stefanprodan/gitops-linkerd

**Reply by frank-bee** [reply]
that is a perfect example, the [gitops-linkerd](https://github.com/stefanprodan/gitops-linkerd)
So if I want to install the same infrastructure into several clusters ( 10, 20 or more), I want to avoid duplication of this part:
https://github.com/stefanprodan/gitops-linkerd/blob/main/clusters/my-cluster/infrastructure.yaml

Would it be smart to put all theses Kustomizations in another Parent Kustomization?
I would also version this complete stack via branch / tag. 
Like this I can install e.g. Infra v1.1 in cluster a-dev, b-qa and c-prod.
And at some point of time install infra v1.2 in a-dev first and after some testing in the other clusters.
(BTW, "a" , "b"... would be tenents and "dev", "qa" environments . All virtual k8s clusters)

**Reply by frank-bee** [reply]
The other option would be to fork from a template repository for each tenant.  one similar to [gitops-linkerd](https://github.com/stefanprodan/gitops-linkerd).
Upgrading the "infrastructure" would mean patching a lot in this case....and it has to be automated

**Reply by stefanprodan** [reply]
Yeap the per-cluster Kustomization that applies the infra items and their GitRepo, can have a patch for `.spec.ref.tag`, so you have different tags across the fleet. I suggest having 2 repos for this, the bootstrap one, where the clusters and these top level Kustomizations are, and an infra repo, where you'll be tagging.

**Reply by frank-bee** [reply]
So you mean a Kustomization can reference a specific branch of a Gitrepository? Is there an example for that? I have seen the git tag only in  flux Gitrepository resources.

**Reply by stefanprodan** [reply]
I haven't said that, what i said is that the top level Kustomization can patch the GitRepo of the infra Kustomizations.

**Reply by frank-bee** [reply]
thanks a lot @stefanprodan for your help.
I think , now I understood your proposal :-)
I even started to created an example for such a "fleet components repo" based on your linkerd example
https://github.com/frank-bee/gitops-fleet-components

**Reply by frank-bee** [reply]
I think the patching can even be avoided, if the top level Kust. and the GitRepository is in the fleet repository .
Fixed : https://github.com/frank-bee/gitops-fleet-components
